flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java

```diff
@@ -0,0 +1,357 @@
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph));
+
+		final int numConcurrentAccesses = 10;
+
+		final ArrayList<CompletableFuture<AccessExecutionGraph>> executionGraphFutures = new ArrayList<>(numConcurrentAccesses);
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			for (int i = 0; i < numConcurrentAccesses; i++) {
+				CompletableFuture<AccessExecutionGraph> executionGraphFuture = CompletableFuture
+					.supplyAsync(
+						() -> executionGraphCache.getExecutionGraph(jobId, jobManagerGateway),
+						TestingUtils.defaultExecutor())
+					.thenCompose(Function.identity());
+
+				executionGraphFutures.add(executionGraphFuture);
+			}
+
+			final CompletableFuture<Collection<AccessExecutionGraph>> allExecutionGraphFutures = FutureUtils.combineAll(executionGraphFutures);
+
+			Collection<AccessExecutionGraph> allExecutionGraphs = allExecutionGraphFutures.get();
+
+			for (AccessExecutionGraph executionGraph : allExecutionGraphs) {
+				assertEquals(accessExecutionGraph, executionGraph);
+			}
+
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId), any(Time.class));
+		}
+	}
+
+	/**
+	 * Tests that a cache entry is invalidated if the retrieved {@link AccessExecutionGraph} is in
+	 * state {@link JobStatus#SUSPENDED}.
+	 *
+	 * <p>This test can be removed once we no longer request the actual {@link ExecutionGraph} from the
+	 * {@link JobManager}.
+	 */
+	@Test
+	public void testCacheInvalidationIfSuspended() throws Exception {
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.hours(1L);
+		final JobID jobId = new JobID();
+
+		final AccessExecutionGraph accessExecutionGraph = mock(AccessExecutionGraph.class);
+
+		final AccessExecutionGraph suspendedExecutionGraph = mock(AccessExecutionGraph.class);
+		when(suspendedExecutionGraph.getState()).thenReturn(JobStatus.SUSPENDED);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		// let's first answer with a JobNotFoundException and then only with the correct result
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(
+			CompletableFuture.completedFuture(suspendedExecutionGraph),
+			CompletableFuture.completedFuture(accessExecutionGraph));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(suspendedExecutionGraph, executionGraphFuture.get());
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
```

Comment:
Need to verify `jobManagerGateway#requestJob` is issued 2 times (to assert cache invalidation).

Reply:
I think this is implicitly tested by having different `AccessExecutionGraphs` returned by the different calls.